### ISSUE TITLE
How does SHAP (TreeExplainer) differ from eli5's approach?

### ISSUE THREAD

**Issue by angeliney** [shap/shap#1466]
I'm not sure if this has been discussed before. I came across eli5's feature importance and found that its values are very highly correlated with SHAP values I got from TreeExplainer (with tree_path_dependent setting). 

In eli5's [documentation](https://eli5.readthedocs.io/en/latest/autodocs/xgboost.html), they referred to this blog as to how they calculated its importance: http://blog.datadive.net/interpreting-random-forests/ 

How similar/different are SHAP (TreeExplainer (tree_path_dependent)) and the approach mentioned in the blog?

**Comment by HughChen** [comment]
My impression is that the eli5 approach is analogous to tree path dependent when you consider only a single ordering of features (which I don't think is bad, but may be a bit simple in my opinion).

SHAP values condition on all possible orderings of features to get at the contribution of a specific feature for all of these orderings.  Tree path dependent approximates the conditional expectation using the samples in the tree as does this eli5 approach.

**Comment by angeliney** [comment]
Thank you for your response! Can you clarify a little bit more on what do you mean when eli5's approach only consider a single ordering of features? What confuses me is based on the blog that eli5 references (http://blog.datadive.net/interpreting-random-forests/), it seems that the ordering does not matter.

**Comment by HughChen** [comment]
My understanding of the eli5 approach is that it takes the path of features (let's say it's x1,x2,x3) along the prediction path and computes the importance of each feature as follows:
\phi_x1 = E[f(X)|x1]-E[f(X)]
\phi_x2 = E[f(X)|x2,x1]-E[f(X)|x1]
\phi_x3 = E[f(X)|x3,x2,x1]-E[f(X)|x1,x2]

It approximates the conditional expectation by using the samples in the tree, so that conditioning on a set of features gives you the average prediction for samples that match the features you are conditioning on.  Interestingly, this is the same way in which tree_path_dependent approximates the conditional expectation.  However, tree_path_dependent doesn't only compute the importance of x1 as above, it computes it for all possible orderings of other features:
\phi_x1 = W1*(E[f(X)|x1]-E[f(X)]) + W2*(E[f(X)|x1,x2]-E[f(X)|x2]) + W3*(E[f(X)|x1,x3]-E[f(X)|x3]) + W4*(E[f(X)|x1,x2,x3]-E[f(X)|x2,x3])
Where Wi are the weighting terms associated with how many orderings correspond to the combination of features we condition on.

Hopefully that helps!  For more detail, feel free to check out the [treeexplainer paper](https://arxiv.org/pdf/1905.04610.pdf).  I think the eli5 method for explaining trees is what we call Saabas in that paper.

**Comment by angeliney** [comment]
I see thank you so much for clarifying! Appreciate it!
